Better binning of interactions

stevenhofmeyr opened the pull request:
This PR changes the way interactions are implemented for work, home, school and nborhood. It uses a single templated function to do most of the work, and for each particular interaction the only requirements are to have a candidate functor (which determines what makes a transmission candidate), a binary interaction function, and to possibly specify a more restrictive binning variable (e.g. workgroup). It also uses `BinPolicy::GPU` to speed up binning.  This should result in significant overall speedups.

stevenhofmeyr committed:
For the work interaction model, bin by workgroup plus community. This results in large speedups, especially for CPU only code. Also, when doing interactions, iterate through the infectious agents first, which is also faster, since most agents will be susceptible and fewer will be infectious. This requires atomic multiply for the prob_ptr for GPUs.

stevenhofmeyr committed:
Updated assert calls to avoid unused variables when not in debug mode.

stevenhofmeyr committed:
Changed code to use PTDType for work bins instead of PType, and removed extra AOS attribute from agents

stevenhofmeyr committed:
Use particle tile data for all binning

stevenhofmeyr committed:
Bin home interactions by nborhood and family cluster

stevenhofmeyr committed:
Enable daytime interactions

stevenhofmeyr committed:
Restructured nborhood binning - because of community interactions can't bin smaller than a community, but doing infectious then susceptible still helps performance

stevenhofmeyr committed:
Cleaned up interaction models

stevenhofmeyr committed:
Switched to GPU policy for building bins. Much faster on GPUs

stevenhofmeyr committed:
Implemented new binner for school interactions

stevenhofmeyr committed:
Trying to fix school interactions - should they be in the same nborhood? Work or home?

stevenhofmeyr committed:
Better matching of school interaction model with development branch

stevenhofmeyr committed:
Restructured interaction model to use a common templated function for work, home, school and nborhood. Much simpler code now.

stevenhofmeyr committed:
Always rebuild the bins - needed since redistribute changes the particle locations

stevenhofmeyr committed:
Fixed indentation

stevenhofmeyr committed:
Removed some unused variables, and fixed reference to SOA index for non-group in nborhood interactions.

stevenhofmeyr committed:
Renamed shadowed variables to avoid PR checks complaints.

stevenhofmeyr committed:
Simplified interaction model template by removing the ACT typename.
Switched binning policy from GPU to Serial, in order to preserve reproducibility for a given random seed.
Commented out the generic interaction model, which is never used.

stevenhofmeyr committed:
Merged in development

stevenhofmeyr committed:
Moved makeBins from InteractionModel to InteractionModRandom because that is the only place it is used and it will be replaced soon.

stevenhofmeyr committed:
Added a parameter for the probability of an agent going on random travel

atmyers commented on src/InteractionModel.H:
```diff
@@ -41,80 +42,178 @@ namespace ExaEpi
+            int max_group = 0;
+            if (bin_idx != -1) {
+                auto bingroup_ptr = soa.GetIntData(bin_idx).data();
+                for (int i = 0; i < np; i++) {
```
This looks like it relies on managed memory to pull the data off the device and run on the host. This could instead use `amrex::ParticleReduce` as here: https://github.com/AMReX-Codes/ExaEpi/blob/development/src/AgentContainer.cpp#L576 to do this calculation on the GPU. 

To turn off managed memory, you can do `amrex.the_arena_is_managed=0`

stevenhofmeyr replied:
Turning off managed memory causes a segfault at `read_workerflow` (line 552) at the call to 
`amrex::The_Arena()->alloc`. Am I correct in thinking that we require managed memory to use this call?

atmyers replied:
Yes, it looks like it does. I can make a fix for that quickly.

stevenhofmeyr replied:
This branch now works fine with managed memory turned off, as long as the `agent.fast=true` parameter is set (that forces use of the GPU binning policy). I see that in main.cpp, line 28, the default is set to managed memory. Can we set this only if not using the fast parameter?

atmyers commented on src/InteractionModel.H:
```diff
@@ -41,80 +42,178 @@ namespace ExaEpi
+            DenseBins<PTDType> bins;
+            // The GPU bin policy is faster, but non-deterministic.
+            // bins.build(BinPolicy::GPU, np, ptd, num_tiles * max_group, binner);
+            bins.build(BinPolicy::Serial, np, ptd, num_tiles * max_group, binner);
```
I think we should evaulate the performance in fast mode, i.e using the `BinPolicy::Gpu` method.

atmyers commented on src/InteractionModel.H:
```diff
@@ -41,80 +42,178 @@ namespace ExaEpi
+            auto& soa = ptile.GetStructOfArrays();
+            auto num_tiles = numTilesInBox(mfi.validbox(), true, bin_size);
+
+            int max_group = 0;
```
I also find that sometimes `max_group` gets set to something huge, I think there is an overflow or something in `CensusData::assignTeachersAndWorkgroup`

stevenhofmeyr replied:
Can you remind me what dataset you used to trigger this?

atmyers replied:
Hi @stevenhofmeyr - it was the `inputs_random_ca` example.

stevenhofmeyr committed:
Removed InteractionModRandom - random travel is now handled in nborhood interactions

stevenhofmeyr committed:
For the interaction models, compute the max group using a parallel reduction, and only do it once.

stevenhofmeyr committed:
Split nborhood interactions into separate models for work and home.
Save the DenseBins for interactions to member of the interaction model. Have an option to reuse without rebuilding. This is disabled at the moment because Redistribute changes the ordering of particles.

stevenhofmeyr committed:
Added command line option agent.fast to switch between fast implementation and slow. Currently this only switches between a serial and a GPU binning policy.

stevenhofmeyr committed:
Removed unused variable

stevenhofmeyr committed:
Fixed issue with double to Real conversion

atmyers committed:
Remove managed memory from read_workerflow

stevenhofmeyr committed:
Fixed overflow when computing number of workgroups for teachers

atmyers committed:
run assignTeachersAndWorkgroup on the host for now

atmyers commented:
I can confirm that with the overflow fix, the bin sizes are reasonable and this can run on my locl GPU again.

stevenhofmeyr committed:
Merged in fix to avoid using managed memory

stevenhofmeyr committed:
Added option for GPU policy binning to CensusData binning

stevenhofmeyr committed:
Removed some unused code and changed a checking loop to not use managed memory.

atmyers commented:
Note that we can tune the `RedistributeGPU` time a lot by changing the `agent.max_grid_size` parameter. At `agent.max_grid_size = 16`, the current default, it takes 280 seconds on a test I ran, but with `agent.max_grid_size=32`, it only took 9. Now that we are calling `Redistribute` much more often we need to change this default.

atmyers commented on src/InteractionModWorkNborhood.H:
```diff
@@ -0,0 +1,102 @@
+            // passing -1 for the binning group indicates bin only by community (cell)
+            interactAgentsImpl<InteractionModWorkNborhood<PCType, PTDType, PType>, PCType, PTDType,
+                                WorkNborhoodCandidate<PTDType>,
+                                BinaryInteractionWorkNborhood<PTDType>>(*this, agents, -1);
```
Why only bin by cell here? When I make separate timers for each type of interaction, it looks like much more time is spent in this one than all the others. Can't we bin this by `IntIdx::work_nborhood`?

atmyers replied:
I see... it's because we also account for general community spread in this routine.

stevenhofmeyr replied:
Yes, this is the slowest interaction, since it's the full community.

stevenhofmeyr commented:
This branch is now superseded by the `fast-interactions` branch.
